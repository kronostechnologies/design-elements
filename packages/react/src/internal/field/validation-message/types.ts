import { NativeHTMLProps } from '../../../types/native-props';

type PartialNativeDivProps = Omit<NativeHTMLProps, 'ref'>;

export interface ValidationMessageProps extends PartialNativeDivProps {
    noInvalidFieldIcon?: boolean;
}
