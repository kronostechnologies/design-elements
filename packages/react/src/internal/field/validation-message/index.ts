export { ValidationMessage } from './validation-message';
