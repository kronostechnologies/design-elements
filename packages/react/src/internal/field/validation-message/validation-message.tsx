import { forwardRef, PropsWithChildren, ReactElement, Ref } from 'react';
import { useDataAttributes } from '../../../hooks/use-data-attributes';
import { useId } from '../../../hooks/use-id';
import { useDeviceContext } from '../../../components/device-context-provider/device-context-provider';
import { useFieldControl } from '../context';
import { StyledIcon, StyledValidationMessage } from './styled';
import { ValidationMessageProps } from './types';

export const ValidationMessage = forwardRef(({
    children,
    id: providedId,
    noInvalidFieldIcon,
    ...otherProps
}: PropsWithChildren<ValidationMessageProps>, ref: Ref<HTMLSpanElement>): ReactElement => {
    const { isMobile } = useDeviceContext();
    const dataAttributes = useDataAttributes(otherProps);
    const { slotIds } = useFieldControl();
    const id = useId(providedId ?? slotIds?.invalid);

    return (
        <StyledValidationMessage
            data-testid="invalid-field"
            aria-live="polite"
            id={id}
            ref={ref}
            role="alert"
            $isMobile={isMobile}
            {...otherProps /* eslint-disable-line react/jsx-props-no-spreading */}
            {...dataAttributes /* eslint-disable-line react/jsx-props-no-spreading */}
        >
            {!noInvalidFieldIcon && (
                <StyledIcon name="alertOctagon" size={isMobile ? '24' : '16'} />
            )}
            {children}
        </StyledValidationMessage>
    );
});

ValidationMessage.displayName = 'InvalidFieldMessage';
