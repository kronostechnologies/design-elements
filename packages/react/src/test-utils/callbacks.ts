export function doNothing(): void {
    // do nothing
}
