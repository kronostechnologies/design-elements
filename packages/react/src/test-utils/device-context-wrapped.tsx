import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
