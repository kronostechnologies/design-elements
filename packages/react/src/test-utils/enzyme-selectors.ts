import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
