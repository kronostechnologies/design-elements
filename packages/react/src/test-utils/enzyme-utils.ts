import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
