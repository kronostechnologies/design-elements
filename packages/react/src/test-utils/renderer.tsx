import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
