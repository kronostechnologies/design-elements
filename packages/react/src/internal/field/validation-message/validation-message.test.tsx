import { renderPortalWithProviders } from '../../../test-utils/renderer';
import { ValidationMessage } from './validation-message';

describe('Validation Message', () => {
    const message = 'Validation Message';
    const customTestId = 'my-test-id';

    describe('Features', () => {
        it('renders correctly with children', () => {
            const { getByText, getByTestId } = renderPortalWithProviders(
                <ValidationMessage>{message}</ValidationMessage>,
            );
            expect(getByText(message)).toBeDefined();
            expect(getByTestId('invalid-field')).toBeDefined();
        });

        test('can override data-testid', () => {
            const { getByTestId } = renderPortalWithProviders(
                <ValidationMessage data-testid={customTestId} />,
            );

            expect(getByTestId(customTestId)).toBeDefined();
        });
    });

    describe('Styling', () => {
        it('matches default snapshot', () => {
            const wrapper = renderPortalWithProviders(
                <ValidationMessage>
                    {message}
                </ValidationMessage>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('matches no icon snapshot', () => {
            const wrapper = renderPortalWithProviders(
                <ValidationMessage noInvalidFieldIcon>
                    {message}
                </ValidationMessage>,
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('matches mobile snapshot', () => {
            const wrapper = renderPortalWithProviders(
                <ValidationMessage>
                    Feedback Message
                </ValidationMessage>,
                'mobile',
            );
            expect(wrapper).toMatchSnapshot();
        });
    });
});
