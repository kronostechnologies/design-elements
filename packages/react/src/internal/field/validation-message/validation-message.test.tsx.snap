// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation Message Styling matches default snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c2 {
  box-sizing: border-box;
  position: fixed;
  z-index: 100000;
  bottom: var(--spacing-2x);
  right: var(--spacing-2x);
}

.c0 {
  color: #CD2C23;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.25rem;
}

.c1 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: var(--spacing-base);
}

<body>
    <div>
      <span
        aria-live="polite"
        class="c0"
        data-testid="invalid-field"
        id="uuid1"
        role="alert"
      >
        <svg
          class="c1"
          color="currentColor"
          focusable="false"
          height="16"
          width="16"
        />
        Validation Message
      </span>
    </div>
    <div
      class="c2"
      data-testid="toasts"
    />
  </body>,
  "container": .c0 {
  color: #CD2C23;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.25rem;
}

.c1 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: var(--spacing-base);
}

<div>
    <span
      aria-live="polite"
      class="c0"
      data-testid="invalid-field"
      id="uuid1"
      role="alert"
    >
      <svg
        class="c1"
        color="currentColor"
        focusable="false"
        height="16"
        width="16"
      />
      Validation Message
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Validation Message Styling matches mobile snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c2 {
  box-sizing: border-box;
  position: fixed;
  z-index: 100000;
  bottom: var(--spacing-2x);
  right: 0;
}

.c0 {
  color: #CD2C23;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.5rem;
}

.c1 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: var(--spacing-base);
}

<body>
    <div>
      <span
        aria-live="polite"
        class="c0"
        data-testid="invalid-field"
        id="uuid1"
        role="alert"
      >
        <svg
          class="c1"
          color="currentColor"
          focusable="false"
          height="24"
          width="24"
        />
        Feedback Message
      </span>
    </div>
    <div
      class="c2"
      data-testid="toasts"
    />
  </body>,
  "container": .c0 {
  color: #CD2C23;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.5rem;
}

.c1 {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: var(--spacing-base);
}

<div>
    <span
      aria-live="polite"
      class="c0"
      data-testid="invalid-field"
      id="uuid1"
      role="alert"
    >
      <svg
        class="c1"
        color="currentColor"
        focusable="false"
        height="24"
        width="24"
      />
      Feedback Message
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Validation Message Styling matches no icon snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c1 {
  box-sizing: border-box;
  position: fixed;
  z-index: 100000;
  bottom: var(--spacing-2x);
  right: var(--spacing-2x);
}

.c0 {
  color: #CD2C23;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.25rem;
}

<body>
    <div>
      <span
        aria-live="polite"
        class="c0"
        data-testid="invalid-field"
        id="uuid1"
        role="alert"
      >
        Validation Message
      </span>
    </div>
    <div
      class="c1"
      data-testid="toasts"
    />
  </body>,
  "container": .c0 {
  color: #CD2C23;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.75rem;
  font-weight: var(--font-normal);
  -webkit-letter-spacing: 0.02rem;
  -moz-letter-spacing: 0.02rem;
  -ms-letter-spacing: 0.02rem;
  letter-spacing: 0.02rem;
  line-height: 1.25rem;
}

<div>
    <span
      aria-live="polite"
      class="c0"
      data-testid="invalid-field"
      id="uuid1"
      role="alert"
    >
      Validation Message
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
